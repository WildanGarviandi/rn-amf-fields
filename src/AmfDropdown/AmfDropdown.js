import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, Text, View, Picker } from 'react-native'
import Icon from 'react-native-vector-icons/MaterialIcons'

class AmfDropdown extends Component {
  constructor(props) {
    super(props)

    this.state = {
      valid: false,
      isFocused: false
    }

  }

  componentDidMount() {
    this.validate()
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.isFocused !== this.state.isFocused) {
      if (!this.state.isFocused) this.validate()
    }
  }

  getLabel = (val) => {
    let item = this.props.item.find(i => i.value === val)
    if (item) return item.label

    return val
  }

  inputValid = () => {
    const { validation, label } = this.props
    if (!validation) return ({status: true})

    // check if pristine
    if (this.props.value === null || this.props.value === undefined) return ({status: true})

    // check if required
    if (validation.required) {
      if (this.props.value === '') return ({status: false, message: label + ' harus dipilih'})
    }

    return ({status: true})
  }

  onDropdownWillShow = () => {
    if (this.props.validation && this.props.validation.required && this.props.value === null) {
      this.props.onSelect('')
    }

    this.setState({isFocused: true})
  }

  onDropdownWillHide = () => {
    this.setState({isFocused: false})
  }

  onSelect = (idx, value) => {
    this.props.onChange(this.props.item[idx].value)
  }

  validate = () => {
    let validationObject = this.inputValid()
    this.setState({valid: validationObject.status, errorMessage: validationObject.message})
  }

  renderSelector() {
    if (this.props.selector) return this.props.selector

    return (
      <View style={[style.option, this.props.optionStyle]}>
        <Text style={style.valueText}>{this.props.value ? this.getLabel(this.props.value) : ''}</Text>
        <Icon name="keyboard-arrow-down" size={16} style={style.icon} />
      </View>
    )
  }

  render() {
    const { layout, value } = this.props

    let validationStyle = {}
    let validationText
    
    if (!this.state.valid) {
      validationStyle = {borderColor: 'red', borderWidth: 1}
      validationText = <Text style={style.errorText}>{this.state.errorMessage}</Text>
    }

    let labelText
    if (this.props.label) {
      labelText = <Text style={[style.label, {marginBottom: layout == 'vertical' ? 8 : 0}]}>{this.props.label}</Text>
    }

    return (
      <View>
        <View style={ layout === 'horizontal' ? style.containerHorizontal : style.container }>
          { labelText }
            <Picker selectedValue={value} onValueChange={this.onSelect} >
              { this.props.list.map(item => <Picker.Item {...item} />) }
            </Picker>
        </View>
        { validationText }
      </View>
    )
  }
}

const style = StyleSheet.create({
  container: {},
  containerHorizontal: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  dropdownContainer: {
    elevation: 4,
    borderRadius: 25,
    borderWidth: 1,
    borderColor: 'transparent',
    backgroundColor: 'white',
    justifyContent: 'center'
  },
  dropdown: {
    width: 100,
    elevation: 4,
    padding: 8,
    height: 150
  },
  errorText: {
    color: 'red',
    margin: 4,
    marginBottom: 0
  },
  label: {
    color: '#454545',
    marginRight: 16
  },
  option: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: 30
  },
  icon: {
    marginRight: 16
  },
  valueText: {
    marginLeft: 16
  }
})

const itemShape = PropTypes.shape({
  label: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired
})

AmfDropdown.defaultProps = {
  layout: 'vertical'
}

AmfDropdown.propTypes = {
  disabled: PropTypes.bool,
  dropdownStyle: PropTypes.object,
  list: PropTypes.arrayOf(itemShape).isRequired,
  label: PropTypes.string,
  layout: PropTypes.oneOf(['vertical', 'horizontal']),
  onChange: PropTypes.func.isRequired,
  optionStyle: PropTypes.object,
  selector: PropTypes.element,
  style: PropTypes.object,
  validation: PropTypes.object,
  value: PropTypes.any,
}

export default AmfDropdown
