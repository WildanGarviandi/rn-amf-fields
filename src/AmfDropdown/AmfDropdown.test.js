import React from 'react'
import { Text } from 'react-native'
import AmfDropdown from './AmfDropdown'
import renderer from 'react-test-renderer'

let list = [{
  label: 'foo',
  value: 'foo'
}, {
  label: 'bar',
  value: 'bar'
}]
let onChange = jest.fn()
let validation = {
  required: true
}

test('AmfDropdown without validation', () => {
  let props = { list, onChange }

  const component = renderer.create(<AmfDropdown {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfDropdown with label', () => {
  let props = { list, onChange, label: 'Label' }
  let propsHorizontal = { ...props, layout: 'horizontal' }

  const component = renderer.create(<AmfDropdown {...props} />)
  const componentHorizontal = renderer.create(<AmfDropdown {...propsHorizontal} />)
  
  let tree = component.toJSON()
  let treeHorizontal = componentHorizontal.toJSON()
  expect(tree).toMatchSnapshot()
  expect(treeHorizontal).toMatchSnapshot()
})

test('AmfDropdown with custom selector', () => {
  let props = {
    list, onChange,
    selector: <Text>Custom Selector</Text>
  }

  const component = renderer.create(<AmfDropdown {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfDropdown with horizontal layout', () => {
  let props = {
    list, onChange,
    layout: 'horizontal'
  }

  const component = renderer.create(<AmfDropdown {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfDropdown with required validation', () => {
  let props = {
    list, onChange, validation
  }
  let propsWithValue = {
    ...props, value: 'foo'
  }

  const component = renderer.create(<AmfDropdown {...props} />)
  const componentWithValue = renderer.create(<AmfDropdown {...propsWithValue} />)
  
  let tree = component.toJSON()
  let treeWithValue = componentWithValue.toJSON()
  expect(tree).toMatchSnapshot()
  expect(treeWithValue).toMatchSnapshot()
})

test('AmfDropdown with unmatch value', () => {
  let props = {
    list, onChange, validation,
    value: 'No Match'
  }

  const component = renderer.create(<AmfDropdown {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfDropdown with empty value', () => {
  let props = {
    list, onChange, validation,
    value: ''
  }

  const component = renderer.create(<AmfDropdown {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfDropdown with empty validation object', () => {
  let props = {
    list, onChange,
    validation: {}
  }

  const component = renderer.create(<AmfDropdown {...props} />)
  
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfDropdown test onChange function', () => {
  let props = {
    list, onChange
  }

  const component = new AmfDropdown(props)
  component.onChange(0, 'value')
  
  expect(props.onChange).toHaveBeenCalled()
})

test('AmfDropdown, test willHide event', () => {
  let props = {
    list, onChange
  }

  const component = new AmfDropdown(props)
  component.setState = jest.fn()
  component.onDropdownWillHide()
  
  expect(component.setState).toHaveBeenCalled()
})

test('AmfDropdown, test willShow event', () => {
  let props = {
    list, onChange
  }
  let propsWithValidation = {
    ...props, validation, value: null
  }

  const component = new AmfDropdown(props)
  component.setState = jest.fn()
  component.onDropdownWillShow()

  const componentWithValidation = new AmfDropdown(propsWithValidation)
  componentWithValidation.setState = jest.fn()
  componentWithValidation.onDropdownWillShow()

  expect(component.setState).toHaveBeenCalled()
  expect(propsWithValidation.onChange).toHaveBeenCalled()
})

test('AmfDropdown, test didUpdate event', () => {
  let props = {
    list, onChange
  }

  const component = new AmfDropdown(props)
  component.state = {isFocused: false}
  component.validate = jest.fn()
  component.componentDidUpdate({}, {isFocused: true})
  expect(component.validate).toHaveBeenCalled()

  const componentFocused = new AmfDropdown(props)
  componentFocused.state = {isFocused: true}
  componentFocused.validate = jest.fn()
  componentFocused.componentDidUpdate({}, {isFocused: false})
  expect(componentFocused.validate).not.toHaveBeenCalled()
})

test('AmfDropdown, test new list', () => {
  let props = {
    list, onChange
  }
  let newItem = [{
    label: 'new',
    value: 'new'
  }]
  let expected = newItem.map(i => i.label)

  const component = new AmfDropdown(props)
  component.componentWillReceiveProps({list: newItem})
  expect(component.list).toEqual(expected)
})
