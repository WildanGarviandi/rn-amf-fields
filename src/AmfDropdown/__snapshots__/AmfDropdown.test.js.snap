// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmfDropdown with custom selector 1`] = `
<View>
  <View
    style={Object {}}
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown with empty validation object 1`] = `
<View>
  <View
    style={Object {}}
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown with empty value 1`] = `
<View>
  <View
    style={Object {}}
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "red",
        "margin": 4,
        "marginBottom": 0,
      }
    }
  >
    undefined harus dipilih
  </Text>
</View>
`;

exports[`AmfDropdown with horizontal layout 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown with label 1`] = `
<View>
  <View
    style={Object {}}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#454545",
            "marginRight": 16,
          },
          Object {
            "marginBottom": 8,
          },
        ]
      }
    >
      Label
    </Text>
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown with label 2`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#454545",
            "marginRight": 16,
          },
          Object {
            "marginBottom": 0,
          },
        ]
      }
    >
      Label
    </Text>
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown with required validation 1`] = `
<View>
  <View
    style={Object {}}
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown with required validation 2`] = `
<View>
  <View
    style={Object {}}
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown with unmatch value 1`] = `
<View>
  <View
    style={Object {}}
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`AmfDropdown without validation 1`] = `
<View>
  <View
    style={Object {}}
  >
    <View
      style={undefined}
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "foo",
              "textColor": undefined,
              "value": "foo",
            },
            Object {
              "label": "bar",
              "textColor": undefined,
              "value": "bar",
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;
